connect/runtime/src/test/java/org/apache/kafka/connect/runtime/isolation/MultiVersionTest.java

```diff
@@ -0,0 +1,296 @@
+            builder.include(VersionedPluginBuilder.VersionedTestPlugin.PREDICATE,
+                DEFAULT_COMBINED_ARTIFACT_VERSIONS.computeIfAbsent(VersionedPluginBuilder.VersionedTestPlugin.PREDICATE, k -> "0.5.0"));
+            DEFAULT_COMBINED_ARTIFACT = Collections.singletonMap(builder.build("all_versioned_artifact"), builder.buildInfos());
+
+            Map<Path, List<VersionedPluginBuilder.BuildInfo>> artifacts = new HashMap<>();
+            artifacts.putAll(DEFAULT_COMBINED_ARTIFACT);
+            artifacts.putAll(DEFAULT_ISOLATED_ARTIFACTS);
+            MULTI_VERSION_PLUGINS = setUpPlugins(artifacts, PluginDiscoveryMode.SERVICE_LOAD);
+
+        } catch (IOException e) {
+            throw new RuntimeException(e);
+        }
+    }
+
+    @Test
+    public void TestVersionedPluginLoaded() throws InvalidVersionSpecificationException, ClassNotFoundException {
+        assertPluginLoad(DEFAULT_COMBINED_ARTIFACT, PluginDiscoveryMode.SERVICE_LOAD);
+        assertPluginLoad(DEFAULT_COMBINED_ARTIFACT, PluginDiscoveryMode.ONLY_SCAN);
+    }
+
+    @Test
+    public void TestMultipleIsolatedVersionedPluginLoading() throws InvalidVersionSpecificationException, ClassNotFoundException {
+        assertPluginLoad(DEFAULT_ISOLATED_ARTIFACTS, PluginDiscoveryMode.SERVICE_LOAD);
+        assertPluginLoad(DEFAULT_ISOLATED_ARTIFACTS, PluginDiscoveryMode.ONLY_SCAN);
+    }
+
+    @Test
+    public void TestLatestVersion() {
+        assertCorrectLatestPluginVersion(DEFAULT_ISOLATED_ARTIFACTS, PluginDiscoveryMode.SERVICE_LOAD, DEFAULT_ISOLATED_ARTIFACTS_LATEST_VERSION);
+        assertCorrectLatestPluginVersion(DEFAULT_ISOLATED_ARTIFACTS, PluginDiscoveryMode.ONLY_SCAN, DEFAULT_ISOLATED_ARTIFACTS_LATEST_VERSION);
+    }
+
+    @Test
+    public void TestBundledPluginLoading() throws InvalidVersionSpecificationException, ClassNotFoundException {
+
+        Plugins plugins = MULTI_VERSION_PLUGINS;
+        // get the connector loader of the combined artifact which includes all plugin types
+        ClassLoader connectorLoader = plugins.pluginLoader(
+            VersionedPluginBuilder.VersionedTestPlugin.SINK_CONNECTOR.className(),
+            PluginUtils.connectorVersionRequirement("0.1.0")
+        );
+        Assertions.assertInstanceOf(PluginClassLoader.class, connectorLoader);
+
+        List<VersionedPluginBuilder.VersionedTestPlugin> pluginTypes = List.of(
+            VersionedPluginBuilder.VersionedTestPlugin.CONVERTER,
+            VersionedPluginBuilder.VersionedTestPlugin.HEADER_CONVERTER,
+            VersionedPluginBuilder.VersionedTestPlugin.TRANSFORMATION,
+            VersionedPluginBuilder.VersionedTestPlugin.PREDICATE
+        );
+        // should match the version used in setUp for creating the combined artifact
+        List<String> versions = Arrays.asList("0.2.0", "0.3.0", "0.4.0", "0.5.0");
```

Comment:
Do these versions correspond to the versions specified in the static initializer? Can they be shared constants?